ocaml/typing/ctype.ml

```diff
@@ -3747,13 +3747,16 @@ exception Filter_arrow_failed of filter_arrow_failure
 
 let filter_arrow env t l ~force_tpoly =
   let function_type level =
-    (* CR layouts v2: This is one of two primary places where we are restricting
-       function arguments / returns to have layout value.  This one handles
+    (* CR layouts v3: This is one of two primary places where we are restricting
```

Comment:
As discussed, I'd like to keep this v2. Maybe we'll backslide on this in a few weeks, but I'm not quite ready to yet.

Reply:
After another discussion, I think the conclusion here is that the questions of (a) what we allow for function args / returns and (b) whether to put that restriction on arrows or functions are orthogonal.  It may turn out that moving the restriction to functions is simple, but I'm not inclined to spend any time on it before we ship floats. 

On the other hand, I do think it's worth trying to allow `any` returns, because otherwise non-returning functions (e.g., any function which is a wrapper around `raise`) can't have a nice type.  I gave this a quick shot just now and predictably ran into various assertion / safety check failures.  I am out of time but will give it a more serious try tomorrow morning.  If there are real complications we can ship without it and follow up after, being slightly sad.

Reply:
OK, I couldn't resist at least looking at an example before running out the door.   Here's a simplified example of a very sad pattern that occurs in the stdlib:

```ocaml
let zero_out a =
  let next i =
    if i = 0 then raise Done else (i-1)
  in
  let rec go i =
    a.(i) := 0;
    go (next i)
  in
  try go (Array.length a) with Done -> ()
```

And for the translation of lambda wants a layout for the return of `go`, so it is sad.  I'll think about this in the morning.

Reply:
In the lambda representation of layouts, we added a layout `Pbottom` for cases such as the result of a raise, where the bottom layout is a sublayout of everything (while it seems to me the layout any is a superlayout of everything?). Wouldn't adding a layout bottom in the typechecker as well work here?

Reply:
Yes, in our lattice "top"/`any` means we don't know anything about the layout and, if we had a bottom, it would mean that we know the layout can be anything. Our "any" is poorly named. I agree non-returning functions can have layout bottom, morally. We can't easily add this layout in the typechecker though, because of how our inference works (for example, we rely on incompatible layouts having no meet).

One observation is that if we finish checking a function and its return layout is any, this means we didn't need to know anything about the layout to check the function.  So the layout doesn't really matter, and could be anything.  From a typing perspective, although I hadn't thought of it this way until now, maybe it would be better to say that the meaning of a fully generalized `'a` of layout `any` is that it could be anything, while the meaning of `any` during inference is that we don't know anything.

After typechecking, any remaining `any`s represent places where no value will really be produced. So maybe we can translate `any` to any layout supported by lambda.  Or maybe we _have_ to pick bottom: otherwise something funny is happening in the `try ... with` - the layouts of the scrutinee and the branch should need to be compatible.  If we, for example, said the return of `go` has layout `void`, they wouldn't be.

Reply:
For what it's worth, I agree with @lpw25 here. `Pbottom` sounds like the correct translation target here.

Reply:
Hmm, I'm not sure the middle/backend is really treating the layouts more like types rather than kinds.  To my mind in these IRs the situation is a slight refinement of what @lpw25 said above: for `go` it's not that it is impossible to call - that's fine, but it _is_ impossible to return from it.  You cannot construct a term that reads the return value because the bottom kind is uninhabited.  Indeed, the one place where `Pbottom` is used in the middle end is in `Cmmgen`, and the semantics is to delete everything following.

The Flambda 2 code does not look at `Pbottom`, so once we're fully switched over it can be removed.

Reply:
I don't think Leo meant impossible to call in the sense you couldn't run the code, but rather impossible to call in just because there's no type you can instantiate `'a` with.

Reply:
On the topic of removing `Pbottom`, I'm not so sure.  We need to pick some (lambda) layout for the return of `go`.  Picking an arbitrary other layout seems odd - if we pick `#float`, for example, we have a `try ... with` where the body had layout `#float` and the branch has layout `immediate`.  It's tempting to pick `void`, but I think then we're still in that situation (oddly incompatible layouts in an expression).

Reply:
It seems to me that since we have a concept of layout subkinding, it would be correct to instantiate `val go ('a : bot). int -> 'a` with any `'a`, since it is only used in covariant positions. Alternatively, while this is forbidden since we have no way to compile it further, `val ignore ('a : any): 'a -> unit` (as a function) could be instantiated with any `'a`, since it is only used in contravariant positions.
Concerning the removal of `Pbottom`, it is needed to be able to construct the layout of an expression, but I think this is again only used in closure and flambda1.

Reply:
> it would be correct to instantiate val go ('a : bot). int -> 'a with any 'a

In general, you can do that trick by instantiating `'a` with an empty type and then saying that the empty type is a subtype of all types of all kinds. Here it's not clear if you can since the empty type is not of kind `bot`, so you can't instantiate it with the empty type in the first place.

> val ignore ('a : any): 'a -> unit (as a function) could be instantiated with any 'a

With that type, `ignore` can already be instantiated with any type since all types have kind `any`. The variance of the type variable is not relevant.

Reply:
In the unboxed type planning meeting just now, we decided to put off allowing `any` returns for a different PR (Mark observed it will require minor flambda2 changes, and anyway it's probably better to do after the PR that properly hooks type system layouts up to lambda layouts).  We still hope to do it for v2.